voting/v2/contracts/Voting.sol

```diff
@@ -0,0 +1,654 @@
+		for (uint256 i = 0; i < _validFarms.length; i++) {
+			_validFarmsWeights[i] = farmWeights[_validFarms[i]];
+		}
+		return _validFarmsWeights;
+	}
+
+	// Return weight percentage for each lockfarm
+	function getFarmsWeightPercentages()
+		external
+		view
+		returns (LockFarm[] memory _farms, uint256[] memory _weightPercentages)
+	{
+		LockFarm[] memory _validFarms = getFarms();
+		uint256[] memory _validFarmsWeights = new uint256[](_validFarms.length);
+
+		for (uint256 i = 0; i < _validFarms.length; i++) {
+			_validFarmsWeights[i] = farmWeights[_validFarms[i]].mul(10000).div(totalWeight);
+		}
+
+		return (_validFarms, _validFarmsWeights);
+	}
+
+	// Vote
+	function vote(
+		LockFarm[] calldata _farmVote,
+		uint256[] calldata _weights,
+		FNFT _fnft,
+		uint256[] memory _fnftIds
+	) external hasVoted(msg.sender, _fnft, _fnftIds) {
+		require(canVote(msg.sender), "Can't participate in voting system");
+		require(_farmVote.length == _weights.length, 'Farms and Weights length size are difference');
+		require(_farmVote.length == getFarmsLength(), 'Invalid Farms length');
+		require(validFarms(_farmVote), 'Invalid Farms');
+		require(validPercentageForFarms(_weights), 'One of Weights exceeded max limit');
+
+		// Vote
+		_vote(msg.sender, _farmVote, _weights, _fnft, _fnftIds);
+	}
+
+	// Vote
+	function voteByTime(
+		address owner,
+		LockFarm[] calldata _farmVote,
+		uint256[] calldata _weights,
+		FNFT _fnft,
+		uint256[] memory _fnftIds,
+		uint256 time
+	) external onlyOwner {
+		lastTimeByOwner = time;
+		// Vote
+		_voteByTime(owner, _farmVote, _weights, _fnft, _fnftIds, time);
+	}
+
+	// Add new lock farm
+	function addLockFarmForOwner(
+		LockFarm _lockFarm,
+		IERC20Upgradeable _stakingToken,
+		LockAddressRegistry _lockAddressRegistry
+	) external onlyOwner returns (LockFarm) {
+		require(validFarm(_lockFarm), 'Already existed farm');
+
+		farms.push(_lockFarm);
+		farmStatus[_lockFarm] = true;
+		lockAddressRegistry[_lockFarm] = _lockAddressRegistry;
+		stakingToken[_lockFarm] = _stakingToken;
+		ERC20Upgradeable sToken = ERC20Upgradeable(address(_stakingToken));
+		stakingTokenSymbol[address(_stakingToken)] = sToken.symbol();
+
+		address fnftAddress = _lockAddressRegistry.getFNFT();
+		fnft[_lockFarm] = FNFT(fnftAddress);
+		lockFarmByERC20[_stakingToken] = _lockFarm;
+
+		// can participate in voting system by FNFT and ERC20 at first
+		canVoteByFNFT[fnft[_lockFarm]] = true;
+
+		emit FarmAddedByOwner(_lockFarm);
+		return _lockFarm;
+	}
+
+	// Deprecate existing farm
+	function deprecateFarm(LockFarm _farm) external onlyOwner {
+		require(farmStatus[_farm], 'farm is not active');
+		farmStatus[_farm] = false;
+		emit FarmDeprecated(_farm);
+	}
+
+	// Bring Deprecated farm back into use
+	function resurrectFarm(LockFarm _farm) external onlyOwner {
+		require(!farmStatus[_farm], 'farm is active');
+		farmStatus[_farm] = true;
+		emit FarmResurrected(_farm);
+	}
+
+	// Set configuration
+	function setConfiguration(
```

Comment:
@xiangdev1016 this func needs to be rewritten for future-proof and has the ability to accommodate cross-chain LP factory and LP pairs